File: import_ledger.py
"""The ledger of what was already imported, so re-runs never duplicate entries.

Every successfully imported thread leaves one line in processed_tweets.txt:

  * an ordinary thread is recorded by its root tweet's ID;
  * a re-imported (extended) thread is additionally recorded as
    "<root id>+<last tweet id>" (see extension_marker), so running the same
    date range twice doesn't import the same extension over and over — while
    a *further* extension later still will be picked up.

Delete the file to import everything from scratch.
"""

from __future__ import annotations

import os
from pathlib import Path
from typing import Optional, Set

from models import Thread

LEDGER_PATH = str(Path(__file__).resolve().parent / "processed_tweets.txt")


def load_processed_tweet_ids() -> Set[str]:
    """Loads the IDs and markers of everything already imported."""
    processed_ids = set()
    if os.path.exists(LEDGER_PATH):
        with open(LEDGER_PATH, "r") as f:
            for line in f:
                processed_ids.add(line.strip())
    return processed_ids


def save_processed_tweet_id(tweet_id: str):
    """Appends one ID (or extension marker) to the ledger file."""
    os.makedirs(os.path.dirname(LEDGER_PATH), exist_ok=True)
    with open(LEDGER_PATH, "a") as f:
        f.write(f"{tweet_id}\n")


def remember_processed(
    tweet_id: str, reimport_marker: Optional[str], processed_tweet_ids: Set[str]
):
    """Records a thread as processed — in the file and in the in-memory set —
    without ever writing the same ID twice."""
    for entry_id in (tweet_id, reimport_marker):
        if entry_id and entry_id not in processed_tweet_ids:
            save_processed_tweet_id(entry_id)
            processed_tweet_ids.add(entry_id)


def extension_marker(thread: Thread) -> str:
    """Identifies a thread together with its current last tweet, so a thread
    that was imported once and extended later can be told apart from the copy
    already in Day One. The '+' keeps these markers distinct from plain tweet
    IDs."""
    return f"{thread[0]['tweet']['id_str']}+{thread[-1]['tweet']['id_str']}"


def delete_ledger():
    """Forgets everything — used by debug mode, which always re-imports its
    tweets and must not leave them recorded afterwards."""
    if os.path.isfile(LEDGER_PATH):
        os.remove(LEDGER_PATH)

File: Twixodus/Pipeline/ImportLedger.swift
// The ledger of what was already imported, so re-runs never duplicate entries.
//
// Every successfully imported thread leaves one line in the ledger file:
//
//   * an ordinary thread is recorded by its root tweet's ID;
//   * a re-imported (extended) thread is additionally recorded as
//     "<root id>+<last tweet id>" (see extensionMarker), so running the same
//     date range twice doesn't import the same extension over and over — while
//     a *further* extension later still will be picked up.
//
// The app keeps one ledger per Twitter account in Application Support, so an
// archive imported today and a fresh archive of the same account imported next
// year share their history — only the new tweets get imported.

import Foundation

public final class ImportLedger {

    /// ~/Library/Application Support/Twixodus
    public static var appSupportFolder: URL {
        FileManager.default
            .urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("Twixodus", isDirectory: true)
    }

    public let fileURL: URL

    /// One ledger per account; the account ID is stable across username
    /// changes. Archives without account.js share the "default" ledger.
    public init(accountId: String?) {
        let name = accountId.flatMap { $0.isEmpty ? nil : $0 } ?? "default"
        fileURL = Self.appSupportFolder
            .appendingPathComponent("ledgers", isDirectory: true)
            .appendingPathComponent("processed_tweets-\(name).txt")
    }

    /// For tests: a ledger at an explicit location.
    public init(fileURL: URL) {
        self.fileURL = fileURL
    }

    /// Loads the IDs and markers of everything already imported.
    public func loadProcessedIDs() -> Set<String> {
        guard let content = try? String(contentsOf: fileURL, encoding: .utf8) else { return [] }
        var ids = Set<String>()
        for line in content.components(separatedBy: .newlines) {
            let trimmed = line.pyStrip()
            if !trimmed.isEmpty {
                ids.insert(trimmed)
            }
        }
        return ids
    }

    /// Appends one ID (or extension marker) to the ledger file. Throws when
    /// the line can't be written — a silently lost line would make the next
    /// run import the same thread again.
    private func append(_ id: String) throws {
        let fm = FileManager.default
        try fm.createDirectory(
            at: fileURL.deletingLastPathComponent(), withIntermediateDirectories: true)

        let data = Data("\(id)\n".utf8)
        // FileHandle only opens existing files; a fresh ledger is created by
        // the write(to:) below instead.
        if let handle = try? FileHandle(forWritingTo: fileURL) {
            defer { try? handle.close() }
            try handle.seekToEnd()
            try handle.write(contentsOf: data)
        } else {
            try data.write(to: fileURL)
        }
    }

    /// Records a thread as processed — in the file and in the in-memory set —
    /// without ever writing the same ID twice. Throws when the ledger file
    /// can't be written; IDs are only added to the set once safely on disk.
    public func rememberProcessed(
        tweetId: String, reimportMarker: String?, processedIDs: inout Set<String>
    ) throws {
        for id in [tweetId, reimportMarker].compactMap({ $0 }) {
            if !processedIDs.contains(id) {
                try append(id)
                processedIDs.insert(id)
            }
        }
    }

    /// Identifies a thread together with its current last tweet, so a thread
    /// that was imported once and extended later can be told apart from the
    /// copy already in Day One. The '+' keeps these markers distinct from
    /// plain tweet IDs.
    public static func extensionMarker(_ thread: TweetThread) -> String {
        "\(thread.first!.idStr)+\(thread.last!.idStr)"
    }
}
